x-pack/test/pki_api_integration/apis/security/pki_auth.ts

```diff
@@ -0,0 +1,370 @@
+
+    expect(cookie.key).to.be('sid');
+    expect(cookie.path).to.be('/');
+    expect(cookie.httpOnly).to.be(true);
+    expect(cookie.maxAge).to.be(0);
+  }
+
+  describe('PKI authentication', () => {
+    before(async () => {
+      await getService('esSupertest')
+        .post('/_security/role_mapping/first_client_pki')
+        .ca(CA_CERT)
+        .send({
+          roles: ['kibana_user'],
+          enabled: true,
+          rules: { field: { dn: 'CN=first_client' } },
+        })
+        .expect(200);
+    });
+
+    it('should reject API requests that use untrusted certificate', async () => {
+      await supertest
+        .get('/api/security/v1/me')
+        .ca(CA_CERT)
+        .pfx(UNTRUSTED_CLIENT_CERT)
+        .set('kbn-xsrf', 'xxx')
+        .expect(401);
+    });
+
+    it('does not prevent basic login', async () => {
+      const [username, password] = config.get('servers.elasticsearch.auth').split(':');
+      const response = await supertest
+        .post('/api/security/v1/login')
+        .ca(CA_CERT)
+        .pfx(UNTRUSTED_CLIENT_CERT)
+        .set('kbn-xsrf', 'xxx')
+        .send({ username, password })
+        .expect(204);
+
+      const cookies = response.headers['set-cookie'];
+      expect(cookies).to.have.length(1);
+
+      const cookie = request.cookie(cookies[0])!;
+      checkCookieIsSet(cookie);
+
+      const { body: user } = await supertest
+        .get('/api/security/v1/me')
+        .ca(CA_CERT)
+        .pfx(UNTRUSTED_CLIENT_CERT)
+        .set('kbn-xsrf', 'xxx')
+        .set('Cookie', cookie.cookieString())
+        .expect(200);
+
+      expect(user.username).to.eql(username);
+      expect(user.authentication_realm).to.eql({ name: 'reserved', type: 'reserved' });
+    });
+
+    it('should properly set cookie and authenticate user', async () => {
+      const response = await supertest
+        .get('/api/security/v1/me')
+        .ca(CA_CERT)
+        .pfx(FIRST_CLIENT_CERT)
+        .expect(200);
+
+      const cookies = response.headers['set-cookie'];
+      expect(cookies).to.have.length(1);
+
+      const sessionCookie = request.cookie(cookies[0])!;
+      checkCookieIsSet(sessionCookie);
+
+      expect(response.body).to.eql({
+        username: 'first_client',
+        roles: ['kibana_user'],
+        full_name: null,
+        email: null,
+        enabled: true,
+        metadata: {
+          pki_delegated_by_realm: 'reserved',
+          pki_delegated_by_user: 'elastic',
+          pki_dn: 'CN=first_client',
+        },
+        authentication_realm: { name: 'pki1', type: 'pki' },
+        lookup_realm: { name: 'pki1', type: 'pki' },
+      });
+
+      // Cookie should be accepted.
+      await supertest
+        .get('/api/security/v1/me')
+        .ca(CA_CERT)
+        .pfx(FIRST_CLIENT_CERT)
+        .set('Cookie', sessionCookie.cookieString())
+        .expect(200);
+    });
+
+    it('should update session if new certificate is provided', async () => {
```

Comment:
__note:__ or be forwarded for `overwritten_session` as I explained in the comment above....